sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncClient.java

```diff
@@ -473,4 +450,77 @@ public CosmosAsyncDatabase getDatabase(String id) {
     public void close() {
         asyncDocumentClient.close();
     }
+
+    TracerProvider getTracerProvider(){
+        return this.tracerProvider;
+    }
+
+    private CosmosPagedFlux<CosmosDatabaseProperties> queryDatabasesInternal(SqlQuerySpec querySpec, CosmosQueryRequestOptions options){
+        return UtilBridgeInternal.createCosmosPagedFlux(pagedFluxOptions -> {
+            String spanName = "queryDatabases";
+            pagedFluxOptions.setTracerInformation(this.tracerProvider, spanName, this.serviceEndpoint, null);
+            setContinuationTokenAndMaxItemCount(pagedFluxOptions, options);
+            return getDocClientWrapper().queryDatabases(querySpec, options)
+                .map(response -> BridgeInternal.createFeedResponse(
+                    ModelBridgeInternal.getCosmosDatabasePropertiesFromV2Results(response.getResults()),
+                    response.getResponseHeaders()));
+        });
+    }
+
+
+    private Mono<CosmosDatabaseResponse> createDatabaseIfNotExistsInternal(CosmosAsyncDatabase database,
+                                                                           ThroughputProperties throughputProperties, Context context) {
+        String spanName = "createDatabaseIfNotExists." + database.getId();
+        Context nestedContext = context.addData(TracerProvider.COSMOS_CALL_DEPTH, TracerProvider.COSMOS_CALL_DEPTH_VAL);
```

Comment:
So this is to avoid nesting tracer call. Initially we were using subscriberContext for this purpose, and found some perf regression. Then we moved to this model, we already have internal methods that are taking context, and eventually we can filter the nesting call in TracerProvide.  If we have taken map, this logic would be more complex in code readability and we have to handle this case by changing code at many places, and also it is more efficient .
And we are using optional when we are reading it back, please check [this](https://github.com/simplynaveen20/azure-sdk-for-java/blob/9c1d31f8b46f43c312fbb3dfcecc3837369d5e08/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/TracerProvider.java#L129)

Reply:
I was just suggesting if we could use a different data structure than the `Context` type here since the Context object is a linked list representation and if its populated enough could result the `getData` to get slower.

Reply:
This is only used for api with nested call , and putting just one data. This is the proper way as nested call will be treated as user send context,  and tracer provider will filter it. This code flow is inline with non nested call, and code readability is also good here.
Also did the perf testing with this approach and numbers looks good. I fear if we change anything we might introduce some perf regression as current approach is working best for us.

Reply:
We will go by this approach,i.e adding nested data only on the api it needed , as this gave the best perf result, and adding data on all api might have an impact .